tentacle/src/transports/tls.rs

```diff
@@ -0,0 +1,254 @@
+    tls_config: Box<ServerConfig>,
+}
+
+impl TlsListener {
+    fn new(timeout: Duration, listen: TcpListener, tls_config: ServerConfig) -> Self {
+        let (sender, rx) = channel(24);
+        TlsListener {
+            inner: listen,
+            timeout,
+            sender,
+            pending_stream: rx,
+            tls_config: Box::new(tls_config),
+        }
+    }
+
+    fn poll_pending(
+        &mut self,
+        cx: &mut Context,
+    ) -> Poll<Option<std::result::Result<(Multiaddr, TlsStream), io::Error>>> {
+        match Pin::new(&mut self.pending_stream).as_mut().poll_next(cx) {
+            Poll::Ready(Some(res)) => Poll::Ready(Some(Ok(res))),
+            Poll::Ready(None) | Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+impl Stream for TlsListener {
+    type Item = std::result::Result<(Multiaddr, TlsStream), io::Error>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        if let Poll::Ready(res) = self.poll_pending(cx) {
+            return Poll::Ready(res);
+        }
+
+        match self.inner.poll_accept(cx)? {
+            Poll::Ready((stream, _)) => match stream.peer_addr() {
+                Ok(remote_address) => {
+                    let timeout = self.timeout;
+                    let mut sender = self.sender.clone();
+                    let acceptor = TlsAcceptor::from(Arc::new(*self.tls_config.clone()));
+                    crate::runtime::spawn(async move {
+                        match crate::runtime::timeout(timeout, acceptor.accept(stream)).await {
+                            Err(_) => warn!("accept tls server stream timeout"),
+                            Ok(res) => match res {
+                                Ok(stream) => {
+                                    let mut addr = socketaddr_to_multiaddr(remote_address);
+                                    addr.push(Protocol::Tls(Cow::Borrowed("/")));
+                                    if sender.send((addr, Box::new(stream))).await.is_err() {
+                                        warn!("receiver closed unexpectedly")
+                                    }
+                                }
+                                Err(err) => {
+                                    warn!("accept tls server stream err: {:?}", err);
+                                }
+                            },
+                        }
+                    });
+                    self.poll_pending(cx)
+                }
+                Err(err) => {
+                    warn!("stream get peer address error: {:?}", err);
+                    Poll::Pending
+                }
+            },
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+/// Tcp transport
+#[derive(Default)]
+pub struct TlsTransport {
+    timeout: Duration,
+    config: TlsConfig,
+    domain_name: String,
+}
+
+impl TlsTransport {
+    pub fn new(timeout: Duration, config: TlsConfig, domain_name: String) -> Self {
```

Comment:
We can get the `domain_name` from the passed parameter `address` in `dial/listen` fn, there is no need to pass it here as a struct field.

Reply:
Because of transport type judge, there get `domain_name` is natural. Of course we could get `domain_name` form `address`, maybe more elegant method is from `TlsTransport` struct which could avoide invoke `find_type` twice.

If have any other comments?  I could modify in one commit. Thanks a lot.

Reply:
From an api design point of view, this is more natural and convenient to call:

```
let transport1 = TlsTransport::new(config);
transport1.listen(addr1);

let transport2 = TlsTransport::new(config);
transport2.dial(addr2);
```

That would look weird:
```
let transport1 = TlsTransport::new(config, extract_domain_name_from(addr1));
transport.listen(addr1);

let transport2 = TlsTransport::new(config, extract_domain_name_from(addr2));
transport2.listen(addr2);
```

Reply:
yeah it looks redundant.
I just thought `TlsTransport::new` only be invoked here:
https://github.com/Pencil-Yao/tentacle/blob/fa33cb3619b495c4455043e592c37cb37fc26e1f/tentacle/src/transports/mod.rs#L194-L210
And the domain name could be got from parsing multiaddress. What's more I thought only in TLS mod could executor here. If have other purposes I don't consider, your pattern is better exactly.

At all, it's ok to modify to your pattern. Wait your reply.

Reply:
@driftluo need your input for this api design.

Reply:
From the point of view of consistency, you can refer to the memory design. Type is only used as a path identifier. The port is actually resolved inside the memory, which means it has been traversed twice. However, I don’t think there is any problem with this traversal, it's the protection of internal information, leaking to the top layer may have discordant places, of course, this does not affect the function 

https://github.com/nervosnetwork/tentacle/blob/master/tentacle/src/transports/memory.rs#L30